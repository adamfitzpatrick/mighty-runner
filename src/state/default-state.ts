import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic, SpecialAttributes } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  specialAttributes: SpecialAttributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  specialAttributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
