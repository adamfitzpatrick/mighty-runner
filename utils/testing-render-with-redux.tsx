import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import * as Models from '@models'
import { AppState } from '../src/state/default-state'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters: (state: Models.Character[] | null = null) => state,
  activeCharacter: (state: Models.CharacterIdentifier | null = null) => state,
  personalData: (state: Models.PersonalData | null = null) => state,
  attributes: (state: Models.Attributes | null = null) => state,
  gear: (state: Models.GearItem[] | null = null) => state,
  effects: (state: Models.Effect[] | null = null) => state
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
