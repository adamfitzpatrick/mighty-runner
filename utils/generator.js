const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
