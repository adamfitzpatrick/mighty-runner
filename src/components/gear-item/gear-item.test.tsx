import * as React from 'react'
import { fireEvent, cleanup, render, RenderResult } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import GearItem, { GearItemEditRender } from '.'
import * as Models from '@models'
import { AppState } from '@state/default-state'
import EditItem from '@components/edit-item'

describe('GearItem', () => {
  let appState: AppState
  let gearItem: Models.GearItem
  let effect: Models.Effect

  beforeEach(() => {
    gearItem = {
      id: '1',
      name: 'n',
      shortName: 's',
      description: 'd',
      asEffectTarget: [ 'a', 'b' ],
      effects: [ '1' ],
      cost: 1,
      availability: '6R',
      value: { rating: 3 }
    }
    effect = {
      id: '1',
      name: 'n',
      description: 'd',
      target: [ 'c', 'd' ],
      value: 1,
      active: true
    }
    appState = {
      gear: [ gearItem ],
      effects: [ effect ]
    } as AppState
  })

  describe('display component', () => {
    let sut: RenderWithRedux
    let editSpy: jest.Mock<Models.GearItem>

    beforeEach(() => {
      editSpy = jest.fn()
      sut = renderWithRedux(<GearItem item={ gearItem } edit={ editSpy } />, appState)
    })

    test('should render correctly', () => {
      expect(sut.container.innerHTML).toMatchSnapshot()
    })

    test('should call the edit function when the edit button is clicked', () => {
      const button = sut.getByTestId('Edit.button.component')
      fireEvent.click(button)
      expect(editSpy).toHaveBeenCalled()
    })
  })

  describe('edit component', () => {
    let sut: RenderResult
    let changeSpy: jest.Mock<Models.GearItem>

    beforeEach(() => {
      cleanup()
      changeSpy = jest.fn()
      sut = render(GearItemEditRender(gearItem, changeSpy))
    })

    test('should render correctly', () => {
      expect(sut.container.innerHTML).toMatchSnapshot()
    })

    test('should call the changeHandler on an input change', () => {
      const nameInput = sut.getByTestId('Name.input.component')
      fireEvent.change(nameInput, { target: { value: 'changed' } })
      gearItem.name = 'changed'
      expect(changeSpy).toHaveBeenCalledWith(gearItem)
    })

    test('should gracefully handle a non-numeric input in a numeric field', () => {
      const costInput = sut.getByTestId('Cost.input.component')
      fireEvent.change(costInput, { target: { value: 'nope' } })
      gearItem.cost = 0
      expect(changeSpy).toHaveBeenCalledWith(gearItem)
    })

    describe('value field', () => {
      let ratingInput: HTMLInputElement

      beforeEach(() => { ratingInput = sut.getByTestId('Rating.input.component') as HTMLInputElement })

      test('should handle a change', () => {
        fireEvent.change(ratingInput, { target: { value: '6' } })
        gearItem.value.rating = 6
        expect(changeSpy).toHaveBeenCalledWith(gearItem)
      })

      test('should gracefully handle a non-numeric input', () => {
        fireEvent.change(ratingInput, { target: { value: 'nope' } })
        gearItem.value.rating = 0
        expect(changeSpy).toHaveBeenCalledWith(gearItem)
      })
    })
  })
})
