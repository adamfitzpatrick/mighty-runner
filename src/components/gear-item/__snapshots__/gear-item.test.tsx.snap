// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GearItem display component should render correctly 1`] = `"<div data-testid=\\"n.gear-item.component\\" class=\\"gearItem\\"><h3>n</h3><div>Description: d</div><div>Cost: 1</div><div>Rating: 3</div><div>Availability: 6R</div><div>Effects:</div><div class=\\"effects\\"><div data-testid=\\"non-editable-effect.component\\"><span>Name: n</span><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"></label></div></div><button class=\\"button\\" data-testid=\\"Edit Gear Item.button.component\\">Edit Gear Item</button><button class=\\"button\\" data-testid=\\"Delete Gear Item.button.component\\">Delete Gear Item</button></div>"`;

exports[`GearItem edit component should render correctly 1`] = `"<div data-testid=\\"1.gear-item-edit-render.component\\"><label><span>Name</span><input data-testid=\\"Name.input.component\\" type=\\"text\\" value=\\"n\\"></label><label><span>Description</span><input data-testid=\\"Description.input.component\\" type=\\"text\\" value=\\"d\\"></label><label><span>Cost</span><input data-testid=\\"Cost.input.component\\" type=\\"text\\" value=\\"1\\"></label><label><span>Availability</span><input data-testid=\\"Availability.input.component\\" type=\\"text\\" value=\\"6R\\"></label><label><span>Rating</span><input data-testid=\\"Rating.input.component\\" type=\\"text\\" value=\\"3\\"></label></div>"`;
